services/code_hosting/git_hub_service.rb

```diff
@@ -1,118 +1,352 @@
+      #  }
+      # @param repo_url [String]
+      # @param branch [String]
+      # @param path [String]
+      # @param provider_credential [GithubProviderCredential]
+      def peek_fastfile_configuration(repo_url: nil, branch: "master", provider_credential: nil, path: self.temp_path, cache: true)
+        repo = repo_from_url(repo_url)
+        return self.cache[repo + branch] if cache && self.cache && !self.cache[repo + branch].nil? && self.cache[repo + branch].kind_of?(Hash)
+        path = File.join(path, repo, branch)
+        begin
+          git_path = File.join(path, repo.split("/").last)
+          # This triggers the check of an existing repo in the given path,
+          # we recover from the error making the clone and checkout
+          Git.open(git_path)
+          fastfile_path = self.fastfile_path(root_path: git_path)
+          fastfile = Fastlane::FastfileParser.new(path: fastfile_path)
+          fastfile_config = {}
+          fastfile.tree.each_key do |key|
+            if key.nil?
+              fastfile_config[:no_platform] = fastfile.tree[key]
+            else
+              fastfile_config[key.to_sym] = fastfile.tree[key]
+            end
+          end
+          fastfile_json_path = File.join(path, "fastfile.json")
+          FileUtils.touch(fastfile_json_path) unless File.exist?(fastfile_json_path) && !File.zero?(fastfile_json_path)
+          File.write(fastfile_json_path, JSON.pretty_generate(fastfile_config))
+          self.cache[repo + branch] = fastfile_config
+          return fastfile_config
+        rescue ArgumentError
+          self.setup_auth(
+            repo_url: repo_url,
+            provider_credential: provider_credential,
+            path: path
+          )
+          self.clone(
+            repo_url: repo_url,
+            branch: branch,
+            provider_credential: provider_credential
+          )
+          self.peek_fastfile_configuration(
+            repo_url: repo_url,
+            branch: branch,
+            provider_credential: provider_credential
+          )
+          self.unset_auth
+        end
+      end
+
+      # Class method that shallow clones the repo on the target branch.
+      # @param [String] repo_url
+      # @param [String] branch
+      # @param [GithubProviderCredential] provider_credential
+      # @return [Git::Base]
+      def clone(repo_url: nil, branch: "master", provider_credential: nil, path: self.temp_path)
+        repo = self.repo_from_url(repo_url)
+        path = File.join(path, repo, branch)
+        FileUtils.rm_rf(path) if File.directory?(path)
+        FileUtils.mkdir_p(path)
+        self.setup_auth(repo_url: repo_url, provider_credential: provider_credential, path: path)
+        Git.clone(url_from_repo(repo_url), repo.split("/").last,
+                  path: path,
+                  recursive: true,
+                  depth: 1)
+        git = Git.open(File.join(path, repo.split("/").last))
+        git.branch(branch).checkout
```

Comment:
Also, @taquitos, correct me if I'm wrong, I'm creating a `/org/repo/branch/` under `File.join(Dir.tmpdir, ".fastlane")` dir, whenever a checkout is being made, the repo is fetched clean from origin, doesn't it? (So multiple peeks to the same repo and branch doesn't duplicate folders)

Reply:
yes, I think that's the best approach. Clean checkouts are the best 💯